Extension/src/LanguageServer/protocolFilter.ts

```diff
@@ -72,47 +69,51 @@ export function createProtocolFilter(): Middleware {
             }
         },
         didChange: async (textDocumentChangeEvent, sendMessage) => {
+            await clients.ActiveClient.awaitUntilLanguageClientReady();
             const me: Client = clients.getClientFor(textDocumentChangeEvent.document.uri);
-            me.onDidChangeTextDocument(textDocumentChangeEvent);
-            sendMessage(textDocumentChangeEvent);
+            if (me.TrackedDocuments.has(textDocumentChangeEvent.document)) {
+                me.onDidChangeTextDocument(textDocumentChangeEvent);
+                await sendMessage(textDocumentChangeEvent);
+            }
         },
-        willSave: defaultHandler,
+        willSave: invoke1,
         willSaveWaitUntil: async (event, sendMessage) => {
```

Comment:
If this handler doesn't use `await` it should not be marked `async`

Reply:
I'm not sure that's necessarily the right rule to apply.  The use of `async` here causes the return type of the lambda to automatically be treated as a `Thenable<TextEdit[]>` without having to explicitly specify that as a return type.  Normally, I would simply use `await` on the call there to `sendMessage`, but that results in a lint warning that the use of `await` is redundant when used in `return await ...`.  If it weren't for the lint error, I would agree.  But, since that occurs, this appears to be the cleanest approach to me.

Reply:
fair enough.